### PR TITLE
fix: discard duplicate cart event

### PR DESCRIPTION
**Pull request by anantjain45823**
## Description of the change

> Discard Duplicate Event for cart events that are generated by us

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/sources/shopify/transform.js [hunk 1]
```diff
@@ -150,6 +163,15 @@ const processEvent = async (inputEvent, metricMetadata) => {
...
       await setAnonymousIdorUserIdFromDb(message, metricMetadata);
+      if (message.anonymousId === "discard") { // cart create event generated by us. Hence discarding
+        return {
+          outputToSource: {
+            body: Buffer.from('OK').toString('base64'),
+            contentType: 'text/plain',
+          },
+          statusCode: 200,
+        }
+      }
```
Comment (utsabc) [inline src/v0/sources/shopify/transform.js:174]: What is this scenario?

Comment (utsabc) [reply src/v0/sources/shopify/transform.js:174]: Why can't we Directly stop allowing `carts_create` events?

Comment (anantjain45823) [reply src/v0/sources/shopify/transform.js:174]: How can we directly stop allowing carts_create event? Like we need some information to know whether it is created by user or not

Comment (yashasvibajpai) [reply src/v0/sources/shopify/transform.js:174]: I suppose we do not support `carts_create` event in Shopify client side but rather all `carts_create` events are generated by us only and internally used. Isn't that the case?

File: src/v0/sources/shopify/config.js [hunk 2]
```diff
@@ -24,7 +24,7 @@ const RUDDER_ECOM_MAP = {
 
 const SHOPIFY_TRACK_MAP = {
   checkouts_delete: 'Checkout Deleted',
-  carts_create: 'Cart Create',
+  // carts_create: 'Cart Create',
```
Comment (yashasvibajpai) [inline src/v0/sources/shopify/config.js:27]: Why is this commented? Is there a later plan to re-introduce it?

File: test/__tests__/data/shopify.json [hunk 3]
```diff
@@ -15,29 +15,7 @@
...
-            "type": "track",
-            "properties": {
-                "created_at": "2023-02-10T12:05:04.402Z",
-                "id": "shopify_test3",
-                "note": null,
-                "products": [],
-                "token": "shopify_test3",
-                "updated_at": "2023-02-10T12:16:07.251Z"
-            }
+            "error": "event type carts_create not supported"
```
Comment (yashasvibajpai) [inline test/__tests__/data/shopify.json:18]: Can we absorb the changes made [here](https://github.com/rudderlabs/rudder-transformer/pull/2082) in this PR ? 
CC @krishna2020

Comment (anantjain45823) [reply test/__tests__/data/shopify.json:18]: I am adding some more cases here and there first it is confirm.Then we can see depending on number conflicts
Will that be right approach?

File: src/v0/sources/shopify/config.js (hunk 2 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/config.js]: Nope. 
Removed

File: src/v0/sources/shopify/transform.js (hunk 1 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/transform.js]: okay . Removed

File: src/util/redisConnector.js [hunk 4]
```diff
@@ -81,23 +98,38 @@ const RedisDB = {
   /**
    * Used to set value in redis depending on the key and the value type
    * @param {*} key key for which value needs to be stored
-   * @param {*} value value to be stored in redis
-   * @param {*} isValJson set to false if value is not a json object
+   * @param {*} value value to be stored in redis send it in array format [field1, value1, field2, value2]
```
Comment (ItsSudip) [inline src/util/redisConnector.js:101]:
```suggestion
   * @param {*} value to be stored in redis send it in array format [field1, value1, field2, value2]
```

File: src/v0/sources/shopify/util.js [hunk 5]
```diff
@@ -190,14 +183,37 @@ const setAnonymousIdorUserIdFromDb = async (message, metricMetadata) => {
...
+const isValidCartEvent = async (inputEvent) => {
+  const cartToken = inputEvent.token || inputEvent.id;
+  const redisVal = await RedisDB.getVal(cartToken);
+  const newCartItems = inputEvent?.line_items.length !== 0 ? sha256(inputEvent.line_items) : 0;
+  if (redisVal) {
+    const prevCartItems = redisVal.itemsHash;
+    if (prevCartItems === newCartItems) {
+      return false;
+    }
+    await anonymousIdCache.get(cartToken, () => redisVal.anonymousId);
```
Comment (ItsSudip) [inline src/v0/sources/shopify/util.js:203]: Shouldn't it be set?

File: src/util/redisConnector.js [hunk 6]
```diff
@@ -59,20 +64,32 @@ const RedisDB = {
...
+      if (isObjExpected === true) {
+        value = await this.client.hmget(key);
+        if (isDefinedAndNotNull(value)) {
+          Object.keys(value).forEach(Key => {
+            try {
+              value[Key] = JSON.parse(value[Key]);
+            } catch (e) {
+              // do nothing
+            }
+          });
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/util/redisConnector.js:80]: The object is stores in this format
```
"key": {
            "key1" : JSON_STRING(""),
            "key2" : JSON_STRING("")
        }
```

File: src/v0/sources/shopify/util.js (hunk 5 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/util.js:203]: its from cache class. We don't have set function but instead we pass a fallback function through which we can set the val. Ref:
https://github.com/rudderlabs/rudder-transformer/blob/383a42a26873e03320d7abbeabaa0dea5856873b/src/v0/util/cache.js#L12

File: src/util/redisConnector.js [hunk 7]
```diff
@@ -59,20 +64,32 @@ const RedisDB = {
    * @returns value which can be json or string or number
    *
    */
-  async getVal(key, isJsonExpected = true) {
+  async getVal(key, isObjExpected = true) {
```
Comment (utsabc) [inline src/util/redisConnector.js:67]: Please update the JS docs its still showing `description` for `isJsonExpected`

File: src/util/redisConnector.js [hunk 8]
```diff
@@ -29,6 +30,10 @@ const RedisDB = {
           if (times <= this.maxRetries) {
             return 10 + times * this.timeAfterRetry;
           }
+          stats.increment('redis_down', {
+            errorType: "Redis",
+            timestamp: Date.now()
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/util/redisConnector.js:35]: Let me gather info and see what would be the best param

File: src/util/redisConnector.js (hunk 7 above)
Comment (anantjain45823) [reply src/util/redisConnector.js:67]: Okay

File: src/util/redisConnector.js [hunk 9]
```diff
@@ -81,23 +98,38 @@ const RedisDB = {
...
+        })
+        bytes = Buffer.byteLength(JSON.stringify(value), "utf-8");
+        await this.client.multi()
+          .hmset(key, ...valueToStore)
+          .expire(key, expiryTime)
+          .exec();
+      } else {
+        bytes = Buffer.byteLength(value, "utf-8");
+        await this.client.setex(key, expiryTime, value);
+      }
```
Comment (utsabc) [inline src/util/redisConnector.js:126]: Remove timestamp as a stats label for all occurrences

File: src/v0/sources/shopify/config.js [hunk 10]
```diff
@@ -102,7 +100,7 @@ const SUPPORTED_TRACK_EVENTS = [
   'orders_paid',
   'orders_partially_fullfilled',
 ];
-const timeDifferenceForCartEvents = 10000; // in micro seconds as we will be compairing it in timestamp 
+const ANONYMOUSID_CACHE_TTL = 60 * 60;
```
Comment (utsabc) [inline src/v0/sources/shopify/config.js:103]:
```suggestion
const ANONYMOUSID_CACHE_TTL_IN_SEC = 60 * 60;
```

File: src/v0/sources/shopify/shopify.util.test.js [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+        it('Customer event -> random AnonymousId', () => {
+            const input = {
+                event: 'Customer Enabled',
+                properties: {
+                    cart_token: '123',
+                }
+            };
+            const output = getAnonymousId(input);
+            expect(output).toEqual(output); // since it will be random
```
Comment (utsabc) [inline src/v0/sources/shopify/shopify.util.test.js:101]: In Customer Enabled event we  don't expect the cart_token field in payload why 
are we adding same in test case?

File: src/v0/sources/shopify/shopify.util.test.js [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        it('Order Delete -> No anonymousId is there', () => {
+            const input = {
+                event: 'Order Deleted',
+                properties: {
+                    cart_token: '123',
+                }
+            };
+            const expectedOutput = null;
+            const output = getAnonymousId(input);
+            expect(output).toEqual(expectedOutput);
```
Comment (utsabc) [inline src/v0/sources/shopify/shopify.util.test.js:113]: Same as above

File: src/v0/sources/shopify/shopify.util.test.js [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+        it('Properties contain id for cart event', async () => {
+            const input = {
+                event: 'Cart Update',
+                properties: {
+                    id: 'shopify_test2',
+                }
+            };
+
+            const expectedOutput = 'anon_shopify_test2';
+            const output = await getAnonymousIdFromDb(input);
```
Comment (utsabc) [inline src/v0/sources/shopify/shopify.util.test.js:138]: Add a test case where we are fetching `anonymousId` from Db but its not there in redis

File: src/util/redisConnector.js [hunk 14]
```diff
@@ -59,20 +64,32 @@ const RedisDB = {
...
+            }
+          });
+        }
+      } else {
+        value = await this.client.get(key);
       }
-      return isJsonExpected ? JSON.parse(value) : value;
+      const bytes = Buffer.byteLength(JSON.stringify(value), "utf-8");
+      stats.gauge('redis_get_val_size', bytes, {
+        timestamp: Date.now()
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/util/redisConnector.js:87]: Done in #2107

File: src/util/redisConnector.js (hunk 9 above)
Comment (anantjain45823) [reply src/util/redisConnector.js:126]: Done in https://github.com/rudderlabs/rudder-transformer/pull/2107

File: src/util/redisConnector.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/util/redisConnector.js:35]: Removed

File: src/util/redisConnector.js [hunk 15]
```diff
@@ -59,20 +64,32 @@ const RedisDB = {
...
+      });
+      return value;
     } catch (e) {
       stats.increment('redis_get_val_error', {
-        error: e,
+        errorType: "Redis",
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/util/redisConnector.js:92]: done

File: src/v0/sources/shopify/shopify.util.test.js (hunk 13 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/shopify.util.test.js:137]: Added

File: src/v0/sources/shopify/shopify.util.test.js (hunk 12 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/shopify.util.test.js]: Corrected

File: src/v0/sources/shopify/shopify.util.test.js (hunk 11 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/shopify.util.test.js]: Corrected

File: src/v0/sources/shopify/config.js (hunk 10 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/config.js]: Done

File: src/util/redisConnector.js [hunk 16]
```diff
@@ -81,23 +98,38 @@ const RedisDB = {
...
+   * @param {*} value to be stored in redis send it in array format [field1, value1, field2, value2]
+   *                  if Value is an object
+   * @param {*} expiryTime expiry time of data in redis by default 1 hr
+   * @param {*} isValJson set to false if value is not a json object 
+   * 
    */
-
-  async setVal(key, value, isValJson = true) {
-    const dataExpiry = 60 * 60; // 1 hour
+  async setVal(key, value, expiryTime = 60 * 60) {
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/util/redisConnector.js]: Done

**Review by utsabc — APPROVED** [review]

File: test/__tests__/data/shopify.json (hunk 3 above)
Comment (krishna2020) [reply test/__tests__/data/shopify.json:18]: Changes are simple:
Move `redisConnector.js` to `src/util/redis` &
Move `redisConnector.test.js` to `src/util/redis` & redis test cases files to `src/util/testdata`

File: src/util/redisConnector.js (hunk 7 above)
Comment (anantjain45823) [reply src/util/redisConnector.js:80]: done

File: src/util/redisConnector.js [hunk 17]
```diff
@@ -72,20 +73,33 @@ const RedisDB = {
...
-      const value = await this.client.get(key);
-      if (value) {
-        const bytes = Buffer.byteLength(value, 'utf-8');
-        stats.gauge('redis_get_val_size', bytes, {
-        });
+      let value;
+      if (isObjExpected === true) {
+        value = await this.client.hmget(key);
+        if (isDefinedAndNotNull(value)) {
+          Object.keys(value).forEach(Key => {
```
Comment (koladilip) [inline src/util/redisConnector.js:87]: Key -> keep key (lower case)

File: src/util/redisConnector.js [hunk 18]
```diff
@@ -122,7 +151,7 @@ const RedisDB = {
       });
       this.client.quit();
```
Comment (koladilip) [inline src/util/redisConnector.js:152]: https://redis.io/commands/quit/ deprecated

File: src/util/redisConnector.js [hunk 19]
```diff
@@ -96,17 +110,32 @@ const RedisDB = {
...
-      await this.client.setex(key, dataExpiry, valueToStore);
+      let bytes;
+      if (typeof value === "object") {
+        const valueToStore = value.map(element => {
+          if (typeof element === "object") {
+            return JSON.stringify(element);
+          }
+          return element;
+        })
+        bytes = Buffer.byteLength(JSON.stringify(value), "utf-8");
```
Comment (koladilip) [inline src/util/redisConnector.js:130]: this is not required, aws provides this already

File: src/util/redisConnector.js [hunk 20]
```diff
@@ -96,17 +110,32 @@ const RedisDB = {
...
+        await this.client.multi()
+          .hmset(key, ...valueToStore)
+          .expire(key, expiryTimeInSec)
+          .exec();
+      } else {
+        bytes = Buffer.byteLength(value, "utf-8");
```
Comment (koladilip) [inline src/util/redisConnector.js:136]: this is not required, aws provides this already

File: src/v0/sources/shopify/util.js [hunk 21]
```diff
@@ -151,38 +136,34 @@ const setAnonymousIdorUserIdFromDb = async (message, metricMetadata) => {
...
       cartToken = properties?.id || properties?.token;
       break;
-    // https://help.shopify.com/en/manual/orders/edit-orders -> order can be edited through shopify-admin only
-    // https://help.shopify.com/en/manual/orders/fulfillment/setting-up-fulfillment -> fullfillments wont include cartToken neither in manual or automatiic
+    // order edit and fullfillments events dont have cart_token or id hence shopify-admin 
     default:
   }
   if (!isDefinedAndNotNull(cartToken)) {
-    stats.increment('shopify_no_cartToken', {
+    stats.increment('No_cartToken_in_payload', {
```
Comment (koladilip) [inline src/v0/sources/shopify/util.js:149]: keep all small no_cart_token_in_payload

what are we planning to do with this metric?

File: src/v0/sources/shopify/util.js [hunk 22]
```diff
@@ -191,26 +172,61 @@ const setAnonymousIdorUserIdFromDb = async (message, metricMetadata) => {
...
+ * @returns true if event is valid else false
+ */
+const isValidCartEvent = async (inputEvent, metricMetadata) => {
+  const cartToken = inputEvent.token || inputEvent.id;
+  let redisVal;
+  try {
+    redisVal = await RedisDB.getVal(cartToken);
+  } catch (e) {
+    // so if redis is down we will send the event to downstream destinations
+    console.log("Error: ", e);
```
Comment (koladilip) [inline src/v0/sources/shopify/util.js:203]: console.log("Shopify events not stitched:", e);

File: test/__tests__/data/shopify.json (hunk 3 above)
Comment (anantjain45823) [reply test/__tests__/data/shopify.json:18]: Okay

File: src/util/redisConnector.js (hunk 17 above)
Comment (anantjain45823) [reply src/util/redisConnector.js:87]: we already have `key` as parameter . will change this to `objKey`

File: src/util/redisConnector.js (hunk 19 above)
Comment (anantjain45823) [reply src/util/redisConnector.js:130]: okay will remove this

File: src/util/redisConnector.js (hunk 20 above)
Comment (anantjain45823) [reply src/util/redisConnector.js:136]: okay

File: src/v0/sources/shopify/util.js (hunk 21 above)
Comment (anantjain45823) [reply src/v0/sources/shopify/util.js:149]: we want to know in how many cases will we wont get any `cartToken` in payload

File: test/__tests__/data/shopify.json (hunk 3 above)
Comment (anantjain45823) [reply test/__tests__/data/shopify.json:18]: done

File: src/util/redisConnector.js (hunk 20 above)
Comment (anantjain45823) [reply src/util/redisConnector.js]: done

File: src/util/redisConnector.js (hunk 19 above)
Comment (anantjain45823) [reply src/util/redisConnector.js]: done

File: src/util/redisConnector.js (hunk 17 above)
Comment (anantjain45823) [reply src/util/redisConnector.js]: done

**Pull request merged**